openoa/utils/downloader.py

```diff
@@ -0,0 +1,428 @@
+    base_url = r"https://goldsmr4.gesdisc.eosdis.nasa.gov/opendap/MERRA2_MONTHLY/M2IMNXLFO.5.12.4/"
+
+    # create save_pathname if it does not exist
+    save_pathname = Path(save_pathname).resolve()
+    if not save_pathname.exists():
+        save_pathname.mkdir()
+
+
+    now = datetime.datetime.now()
+    years = list(range(2000,now.year+1,1))
+
+    
+    
+    # download the data
+    for year in years:
+        
+        # get the file names from the GES DISC site for the year
+        result = requests.get(f"{base_url}{year}")
+        
+        files = re.findall(r"(>MERRA2_\S+.nc4)", result.text)
+        files = list(dict.fromkeys(files))
+        files = [x[1:] for x in files]
+        
+        
+        # coordinate indexes
+        lat_i = ""
+        lon_i = ""
+        
+        # download each of the files and save them
+        for file in files:
+            
+            outfile = save_pathname / f"{save_filename}_{file.split('.')[-2]}.nc"
+
+            if not outfile.is_file():
+
+                # download one file for determining coordinate indicies
+                if lat_i=="":
+                    url = f"{base_url}{year}//{file}"+r".nc4?PS,SPEEDLML,TLML,time,lat,lon"
+                    download_file(url,outfile)
+                    ds_nc = xr.open_dataset(outfile)
+                    ds_nc_idx = ds_nc.assign_coords(lon_idx=("lon",range(ds_nc.dims['lon'])),
+                                                    lat_idx=("lat",range(ds_nc.dims['lat'])))
+                    sel = ds_nc_idx.sel(lat=lat,lon=lon, method="nearest")
+                    lon_i = f"[{sel.lon_idx.values-1}:{sel.lon_idx.values+1}]"
+                    lat_i = f"[{sel.lat_idx.values-1}:{sel.lat_idx.values+1}]"
+                    ds_nc.close()
+                    outfile.unlink() 
+                    
+                # download file with specified coordinates    
+                url = (
+                    f"{base_url}{year}//{file}"
+                    r".nc4?PS[0:0]"
+                    f"{lat_i}{lon_i}"
+                    f",SPEEDLML[0:0]{lat_i}{lon_i}"
+                    f",TLML[0:0]{lat_i}{lon_i}"
+                    f",time,lat{lat_i},lon{lon_i}"
+                )
+
+                download_file(url,outfile)
+                
+
+    # get the saved data
+    ds_nc = xr.open_mfdataset(f"{save_pathname / f'{save_filename}*.nc'}")
+
+    # renamce variables to conform with OpenOA
+    ds_nc = ds_nc.rename_vars({"SPEEDLML":"windspeed_ms","TLML":"temperature_K","PS":"surf_pres_Pa"})
+    
+    # select the central node only for now
+    sel = ds_nc.sel(lat=lat,lon=lon, method="nearest")   
+
+    # convert to a pandas dataframe
+    df = sel.to_dataframe()
+
+    # select required columns
+    df = df[["windspeed_ms","temperature_K","surf_pres_Pa"]]
+
+    # rename the index to match other datasets
+    df.index.name = "datetime"
+
+    # drop any empty rows
+    df = df.dropna()
+    
+    if save_filename is not None:
```

Comment:
Same comment about describing this behavior in the docstring.

Reply:
Done